lsp/ocamllsp.lua

```diff
@@ -20,17 +18,30 @@ local language_id_of = {
   dune = 'dune',
 }
 
-local get_language_id = function(_, ftype)
-  return language_id_of[ftype]
+local language_id_of_ext = {
+  mll = language_id_of.ocamllex,
+  mly = language_id_of.menhir,
+  mli = language_id_of.ocamlinterface,
+}
+
+local get_language_id = function(bufnr, ftype)
+  if ftype == 'ocaml' then
+    local path = vim.api.nvim_buf_get_name(bufnr)
+    local ext = vim.fn.fnamemodify(path, ':e')
+    return language_id_of_ext[ext] or language_id_of.ocaml
+  else
+    return language_id_of[ftype]
+  end
 end
 
 ---@type vim.lsp.Config
 return {
   cmd = { 'ocamllsp' },
   filetypes = { 'ocaml', 'menhir', 'ocamlinterface', 'ocamllex', 'reason', 'dune' },
-  root_dir = function(bufnr, on_dir)
-    local fname = vim.api.nvim_buf_get_name(bufnr)
-    on_dir(util.root_pattern('*.opam', 'esy.json', 'package.json', '.git', 'dune-project', 'dune-workspace')(fname))
-  end,
+  root_markers = {
+    { 'dune-project', 'dune-workspace' },
+    { '*.opam', 'esy.json', 'package.json' },
+    '.git',
+  },
```

Comment:
The "nested" format requires Nvim 0.11.4, hopefully we don't have any ocaml users on older Nvim :)

Reply:
Updated, probably best to keep backwards compat with older NeoVim releases for now.